### PR TITLE
update approval_id before process add trade

### PR DESCRIPTION
**Pull request by cryingraven**

### CODE REVIEW

File: paras-marketplace-contract/src/nft_callbacks.rs [hunk 1]
```diff
@@ -147,23 +147,20 @@ impl NonFungibleTokenApprovalsReceiver for Contract {
...
-                signer_storage_required / storage_amount,
-                storage_amount,
-            );
-
-            let contract_and_token_id = format!("{}{}{}", nft_contract_id, DELIMETER, token_id);
-            if let Some(mut market_data) = self.market.get(&contract_and_token_id) {
-                market_data.approval_id = approval_id;
-                self.market.insert(&contract_and_token_id, &market_data);
+            if owner_paid_storage >= signer_storage_required {
+                env::log_str("Insufficient storage");
```
Comment (emarai) [inline paras-marketplace-contract/src/nft_callbacks.rs:162]: Pakai logging-nya yang "Insufficient storage paid: {}, for {} sales at {} rate of per sale" 

supaya lebih verbose

File: paras-marketplace-contract/src/nft_callbacks.rs [hunk 2]
```diff
@@ -147,23 +147,20 @@ impl NonFungibleTokenApprovalsReceiver for Contract {
...
-                owner_paid_storage,
-                signer_storage_required / storage_amount,
-                storage_amount,
-            );
-
-            let contract_and_token_id = format!("{}{}{}", nft_contract_id, DELIMETER, token_id);
-            if let Some(mut market_data) = self.market.get(&contract_and_token_id) {
-                market_data.approval_id = approval_id;
-                self.market.insert(&contract_and_token_id, &market_data);
+            if owner_paid_storage >= signer_storage_required {
```
Comment (emarai) [inline paras-marketplace-contract/src/nft_callbacks.rs:161]: sudah di-test apa return-nya keexecute?

File: tests/simulation_tests.rs [hunk 3]
```diff
@@ -102,7 +102,206 @@ fn test_add_market_data() {
...
+        outcome.gas_burnt().0 as f64 / 1e12
+    );
+    outcome.assert_success();
+    println!(
+        "[BUY] Restored storage price : {} Bytes",
+        restored_storage_price_for_buy
+    );
+    let expected_gas_ceiling = 50 * u64::pow(10, 12);
+    assert!(outcome.gas_burnt() < Gas(expected_gas_ceiling));
+}
```
Comment (emarai) [inline tests/simulation_tests.rs:304]: Tambahin 2 case lagi: 

# Case 1 
1. user deposit storage
2. user A add trade NFT 1 with NFT 2
3. user A doesn't deposit storage
4. user A add trade NFT 1 with NFT 3
5. other user check that trade NFT 1 with NFT 3 does not exist
6. user B accept trade NFT 2 with NFT 1 (from step 2)

# Case 2
1. user deposit storage
2. user A add trade NFT 1 with NFT 2
3. user A doesn't deposit storage
4. user A add sale with NFT 1 
5. other user check that sale NFT 1 does not exist
6. user B accept trade NFT 2 with NFT 1 (from step 2)

File: tests/simulation_tests.rs [hunk 4]
```diff
@@ -102,7 +102,206 @@ fn test_add_market_data() {
...
+#[test]
+fn test_sale_with_fail_add_trade_storage(){
+    /*
+    1. user deposit storage
+    2. user add market sale
+    3. user doesn't deposit storage
+    4. other user buy the nft
```
Comment (emarai) [inline tests/simulation_tests.rs:216]: tambahin satu step lagi sebelum step 4:
1. user deposit storage
2. user add market sale
3. user doesn't deposit storage
4. user add trade
5. other user check that trade from step 4 does not exist
6. other user buy the nft

File: paras-marketplace-contract/src/nft_callbacks.rs (hunk 2 above)
Comment (cryingraven) [reply paras-marketplace-contract/src/nft_callbacks.rs:161]: aman ke return

**Pull request closed without merging**